Refactor logic analyzer

bessman opened the pull request:
This is the first part of the major refactorization I'm currently working on. It refactors the logic analyzer instrument, as well as the underlying hardware drivers it depends on, including:

- Input Capture (IC)
- Direct Memory Access (DMA)
- Input Change Notification (CN)
- Timers (TMR)

This pull request breaks any instrument which depend on an interrupt service routine from any of the above drivers. This is because I'm refactoring the interrupts to be more modular by using callbacks. The broken instruments will be fixed as part of the ongoing refactorization.

The new low-level hardware drivers currently reside in src/registers_ng, to be renamed to src/registers once the refactorization is complete. These drivers are designed to be self-contained, i.e. they have no dependencies on any resources other than themselves. This design aims to solve the current spaghetti-ness of the firmware code, where every module depends on several other modules, often in a transitive or circular manner.

Depends on #165 and #166.

## Summary by Sourcery

Refactor the logic analyzer instrument and its underlying hardware drivers to improve modularity and reduce code complexity

New Features:
- Introduced a new modular design for logic analyzer hardware drivers in src/registers_ng

Enhancements:
- Redesigned low-level hardware drivers to be self-contained
- Simplified interrupt handling for logic analyzer
- Improved code organization by separating concerns

Chores:
- Removed multiple instrument implementations
- Consolidated hardware driver code
- Updated build system configuration

bessman committed:
Add common types

bessman committed:
Refactor Input Capture driver

bessman committed:
Refactor Direct Memory Access driver

bessman committed:
Refactor pin initialization

bessman committed:
Refactor Input Change Notification driver

bessman committed:
Refactor Timers driver

bessman committed:
Refactor Logic Analyzer

CloudyPadmal commented:
I was thinking about having a separate `development` branch. That way, we have a protected master or main branch and once development is tested, we can merge the two. What do you think?

bessman commented:
In general, I don't think it's necessary for small projects with few committers to have parallel `main` and `development` branches. I prefer to have all development go into `main`, while tagging periodic releases.

On the other hand, merging known broken code (which this is) into `main` is nasty. So in this case, it makes sense to keep this work in a separate branch until the refactorization is complete.

bessman commented:
Going to merge this soon (into the new and probably temporary 'develop' branch).

Some notes on this change:
- Some functionality has been removed from the logic analyzer, specifically:
  - The ability to cascade and prescale timers. This feature is useful to look at signals with long time periods between edges. I see this as a niche use case, but the feature could be re-added in the future.
  - The ability to capture such events as 'every fours rising edge' and 'every sixteenth rising edge' and the like. I don't see a use case for this. Available capture events are now 'any edge', 'falling edge', and 'rising edge'.
  - The ability to configure multiple pins as trigger source simultaneously. I don't see a use case for this. At most one pin at a time can now be configured as trigger source.

bessman commented on src/instruments/logic_analyzer.c:
```diff
@@ -0,0 +1,395 @@
+     */
+    /* Unroll loop; saving even a single clock cycle between DMA channel
+     * starts is meaningful. */
+    switch (g_n_channels) {
+    case 4: // NOLINT(readability-magic-numbers)
+        DMA_start(CHANNEL_4);
+    case 3: // NOLINT(readability-magic-numbers)
+        DMA_start(CHANNEL_3);
+    case 2:
+        DMA_start(CHANNEL_2);
+    case 1:
+        DMA_start(CHANNEL_1);
+    default:
+        break;
+    }
```
The comment immediately preceding the switch statement explains that it is an unrolled loop. Further comments marking the fallthroughs as intentional are redundant.

bessman committed:
fixup

bessman committed:
fixup

bessman committed:
Remove broken modules

There are transitive and circular dependencies everywhere. This began with
removing src/helpers/interval.*, which is superceded by the new logic analyzer
implementation. But the old IC driver transitivelly depended on interval, so
it had to go. Then a bunch of things that depended on IC had to go, and so on
and so forth.

Basically, better to start fresh at this point.

bessman commented on src/registers_ng/dma.c:
```diff
@@ -0,0 +1,481 @@
+    return regs[channel];
+}
+
+static void interrupt_enable(Channel const channel, bool const ctrl)
```
Members of IEC0 and IFS0 structs (DMA0, DMA1, DMA2, DMA3) are bitfields which do not have addresses. Modify your suggestion accordingly.

bessman replied:
Elaborate on the second approach, "switch statement with computed goto or function pointer table".

bessman committed:
Use jump tables for interrupt control

bessman commented:
@sourcery-ai review

bessman commented on src/instruments/logic_analyzer.c:
```diff
@@ -0,0 +1,398 @@
+/* Static Functions */
+/********************/
+
+static void trigger(void)
```
This part of the code is extremely timing critical. Profiling confirms that the switch-case saves clock cycles compared to a for-loop. Gotta go fast.

bessman committed:
Use unified register tables instead of getter function

bessman committed:
Fix shift direction

bessman commented:
@sourcery-ai review

bessman commented on src/instruments/CMakeLists.txt:
```diff
@@ -1,12 +1,7 @@
 target_sources(pslab-firmware.elf PRIVATE
-  logicanalyzer.c
-  multimeter.c
-  oscilloscope.c
+  logic_analyzer.c
   powersource.c
-  sensors.c
-  wavegenerator.c
```
They will be added back in later PRs. This PR temporarily removes them because the low level changes breaks several old implementations.

bessman commented on src/CMakeLists.txt:
```diff
@@ -3,10 +3,17 @@ add_subdirectory(bus)
 add_subdirectory(helpers)
 add_subdirectory(instruments)
 add_subdirectory(registers)
+add_subdirectory(registers_ng)
```
registers_ng will replace the old registers directory when refactorization is complete.

bessman commented on CMakeLists.txt:
```diff
@@ -1,6 +1,7 @@
 cmake_minimum_required(VERSION 3.16)
 set(CMAKE_TOOLCHAIN_FILE external/cmake-microchip/toolchain.cmake)
 set(MICROCHIP_MCU PIC24EP256GP204)
+set(CMAKE_SYSTEM_NAME Generic)
```
CMAKE_SYSTEM_NAME refers to the target operating system. If unset, it defaults to the build host's operating system, which is incorrect in our case. 'Generic' is the correct value for bare metal embedded systems, per cmake's docs.

bessman committed:
Check valid channel value in public functions

bessman committed:
Document LA pin values

bessman committed:
Move enum Status definition to types.h

bessman commented on src/registers_ng/dma.c:
```diff
@@ -0,0 +1,461 @@
+        .p_iec = &IEC2,
+        .p_ifs = &IFS2,
+        .interrupt_mask = DMA3_INTERRUPT_MASK,
+    }
```
Considering this is an embedded system, is assert appropriate? On fail, the system will simply be reset, and the assert error message lost.

bessman replied:
Adding error handling and logging is out of scope for the current PR. I have added input argument checks in all public functions checking for valid channel values. In the future, adding a logging mechanism can be considered.
